src/core/lib/event_engine/posix_engine/file_descriptor_collection.h

```diff
@@ -0,0 +1,207 @@
+// Copyright 2025 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GRPC_SRC_CORE_LIB_EVENT_ENGINE_POSIX_ENGINE_FILE_DESCRIPTOR_COLLECTION_H
+#define GRPC_SRC_CORE_LIB_EVENT_ENGINE_POSIX_ENGINE_FILE_DESCRIPTOR_COLLECTION_H
+
+#include <atomic>
+#include <unordered_set>
+
+#include "absl/status/status.h"
+#include "absl/strings/str_format.h"
+#include "src/core/util/sync.h"
+
+namespace grpc_event_engine::experimental {
+
+class FileDescriptor {
+ public:
+  constexpr FileDescriptor() = default;
+#if GRPC_ENABLE_FORK_SUPPORT
+  constexpr FileDescriptor(int fd, int generation)
+      : fd_(fd), generation_(generation) {};
+#else   // GRPC_ENABLE_FORK_SUPPORT
+  constexpr FileDescriptor(int fd, int /* generation */) : fd_(fd) {};
+#endif  // GRPC_ENABLE_FORK_SUPPORT
+
+  bool ready() const { return fd_ > 0; }
+  // Escape for iomgr and tests. Not to be used elsewhere
+  int iomgr_fd() const { return fd_; }
+  // For logging/debug purposes - may include generation in the future, do not
+  // rely on it for Posix calls!
+  int debug_fd() const { return fd_; }
+
+  constexpr static FileDescriptor Invalid() { return FileDescriptor(-1, 0); }
+
+#if GRPC_ENABLE_FORK_SUPPORT
+  int generation() const { return generation_; }
+  template <typename Sink>
+  friend void AbslStringify(Sink& sink, FileDescriptor fd) {
+    sink.Append(
+        absl::StrFormat("FD(%d), generation: %d", fd.fd_, fd.generation_));
+  }
+#else   // GRPC_ENABLE_FORK_SUPPORT
+  int generation() const { return 0; }
+  template <typename Sink>
+  friend void AbslStringify(Sink& sink, FileDescriptor fd) {
+    sink.Append(absl::StrFormat("FD(%d)", fd.fd_));
+  }
+#endif  // GRPC_ENABLE_FORK_SUPPORT
+
+ private:
+  int fd() const { return fd_; }
```

Comment:
Private accessor methods are not useful, because anything that can access the accessor method can also just directly use the data member.